debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/events/BufferingChangeStreamCursor.java

```diff
@@ -268,6 +273,15 @@ private Optional<ResumableChangeStreamEvent<TResult>> fetchEvent(MongoChangeStre
                     .or(() -> Optional.ofNullable(cursor.getResumeToken()).map(ResumableChangeStreamEvent::new));
         }
 
+        private void waitWhenStreamingPaused(ChangeEventSource.ChangeEventSourceContext context) throws InterruptedException {
+            if (context.isPaused()) {
+                LOGGER.info("Buffering change stream cursor paused");
+                context.streamingPaused();
+                context.waitSnapshotCompletion();
+                LOGGER.info("Buffering change stream cursor resumed");
+            }
```

Comment:
@ani-sha (cc @Naros) I'm not really sure whether this is the right approach either. 

The reason why I liked the initial pause/resume API on the `BufferingChangeStreamCursor` was that the flow of operation remained explicit and clear. 

It allowed for this workflow

The `MongoDbStreamingChangeEventSource` checked the context, 

1) call to `cursor.pause()` (which in turn should pause the fetcher)
2) call `context.streamingPaused();` to notify that the streaming loop reached a safe waiting spot and thus the streaming is to be considered safely paused
3) await snapshot completion by calling `context.waitSnapshotCompletion()`
4) call `cursor.resume()` (which in turn should resume the fetcher)
5) return and continue with the streaming loop 


Added benefit was that there was not tight coupling between `ChangeEventSourceContext` and the `BufferingChangeStreamCursor`. 

having this exact logic in both `BufferingChangeStreamCursor` and `MongoDbStreamingChangeEventSource` makes the action flow quite confusing as it's not clear when exactly is the notification "streaming safely paused, waiting conditions can proceed" (call to `context.streamingPaused()`) invoked -- depending on timing it could be invoked in the fetcher or when `MongoDbStreamingChangeEventSource#waitWhenStreamingPaused` is called. There are two problems tied to this 

1) One of the two calls of `context.streamingPaused()` will redundant (probably safe, however I'd say that the duplicit notification is error prone)
2) Although the call `MongoDbStreamingChangeEventSource#waitWhenStreamingPaused`  is (and should be!) the first operation in the main streaming loop, the trigger inside fetcher is not visible in `MongoDbStreamingChangeEventSource` and could lead a future developer to think that prior to this the streaming is still running.

Reply:
Thus if we decide that creating the aforementioned coupling is OK, the entire "pause" logic should be moved to the `BufferingChangeStreamCursor` and `MongoDbStreamingChangeEventSource#waitWhenStreamingPaused` should be removed, replacing the call to it with a call to `BufferingChangeStreamCursor#waitWhenStreamingPaused` (which would essentially perform the procedure outlined in the previous comment). 

We would have to make sure that potential interrupts are handled correctly

Reply:
> Added benefit was that there was not tight coupling between ChangeEventSourceContext and the BufferingChangeStreamCursor

@jcechace  I agree on this. 
I do recall you suggested this approach in one of yours comments which I can't seem to find this PR now. I will wait for Chris's review and rework accordingly.

Reply:
I think it was in chat, I said it was a possibility which has some drawbacks :)